diff -ru MISP.orig/app/Lib/cakephp/lib/Cake/bootstrap.php MISP/app/Lib/cakephp/lib/Cake/bootstrap.php
--- MISP.orig/app/Lib/cakephp/lib/Cake/bootstrap.php	2020-04-20 21:50:21.000000000 +0000
+++ MISP/app/Lib/cakephp/lib/Cake/bootstrap.php	2020-04-24 17:35:06.155497000 +0000
@@ -155,6 +155,11 @@
  * Full URL prefix
  */
 if (!defined('FULL_BASE_URL')) {
+	if (Configure::read('App.fullBaseUrl')) {
+		define('FULL_BASE_URL', Configure::read('App.fullBaseUrl'));
+	}
+}
+if (!defined('FULL_BASE_URL')) {
 	$s = null;
 	if (env('HTTPS')) {
 		$s = 's';
